accepted/wasm-gc.md

```diff
@@ -0,0 +1,1319 @@
+* [ ] Implement Wasmtime's public API. This is necessary for getting the spec
+      test suite running.
+* [ ] Actually get the spec tests passing in Wasmtime and everything that
+      entails: compilation, runtime support, etc.
+
+From there, to get a production-ready implementation we will need:
+
+* [ ] Wasm GC support in `wasm-encoder`
+* [ ] Wasm GC support in `wasm-smith`
+* [ ] The dedicated fuzzer for exercising Wasm GC
+* [ ] The pluggable GC interface
+* [ ] The general-purpose GC, built with an off-the-shelf collector from MMTK
+* [ ] The semi-space copying collector with tight pooling-allocator integration
+
+# Rationale and Alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+<!-- What other designs have been considered and what is the rationale for chosing -->
+<!-- the one proposed? -->
+
+* We could make all of Wasmtime's public API managed by a global GC. That lets
+  us avoid questions about cyclic type definitions and the lifetimes of
+  canonicalized types in the registry in the engine and what happens when you
+  try to get info about a type from a module that has since been
+  dropped. However, this is a much bigger change to the public API, forces GC on
+  all Wasmtime embedders even if they don't want the Wasm GC proposal, and
+  requires a multi-threaded GC (which this proposal does not). The benefits do
+  not outweight the drawbacks.
+
+* Instead of focusing on a 100% safe, "dynamic" public API, we could focus on
+  getting a "typed" API working. By analogy, we have `wasmtime::Func` and
+  `wasmtime::TypedFunc<...>`, but this proposal is focused only on the former
+  version of exposing GC-managed objects in the public API. I think these two
+  paths are largely orthogonal, but we only have so many engineers to throw at
+  this problem, so I think we have to sequence them. Of the two paths, I think
+  first implementing the "dynamic" version that works for all types we could be
+  given at runtime is the best choice.
+
+* Instead of defining a pluggable GC interface, we could just choose a single
+  implementation and try and make it general-purpose enough to work for all
+  workloads and fit tightly within the pooling allocator's constraints. Doing
+  both at the same time seems like an exceedingly hard engineering problem.
+
+# Open Questions
+[open-questions]: #open-questions
+
+<!-- - What parts of the design do you expect to resolve through the RFC process -->
+<!--   before this gets merged? -->
+<!-- - What parts of the design do you expect to resolve through implementation after -->
+<!--   the RFC is accepted? -->
+
+* When attempting to access information about an array, struct, or function type
+  whose sole defining module has been dropped, should we
+
+  * panic, or
+
+  * return `None` (meaning that all these type accessors return `Option`s)
+
+  ?
+
+* Should `AnyRef` and `EqRef` be collapsed together in the public API? Right now
```

Comment:
If the embedder so chooses, it is possible for internalized external references to inhabit `anyref` without inhabiting `eqref`. For example, in the JS embedding, most internalized references to JS objects are not `eqref`. Something similar might make sense for wasmtime.

More generally, how do you plan to implement `extern.externalize` and `extern.internalize`? I didn’t see that discussed in the RFC.

Reply:
Ah, this is a great point, thanks @tlively. I overlooked those instructions and their implications. I think they do force our hand in two ways:

1. We need to manage `externref`s and `anyref`s with the same collector, since it will be trivial to create cycles between the two, since `externref`s aren't solely created by the host anymore.
2. We do need `wasmtime::AnyRef` that is separate from `wasmtime::EqRef` so that we can represent host-created `externref`s that have been internalized and which do not support equality comparison.

Thanks for helping clear these things up!

As for how to implement the instructions, everything will probably be easier if we adopt a uniform representation between `anyref` and `externref`, but I haven't had a chance to give the problem too much thought yet :)

Reply:
@tlively I added some notes in the most recent commit about these instructions, if you're interested in my current thinking on them.

Reply:
Thanks! Looks good to me.